src/main/php/Sniffs/Rdss/Standards/ZooRoyal/Sniffs/Safe/CheckSafeFunctionUsageSniff.php

```diff
@@ -0,0 +1,169 @@
+use Zooroyal\CodingStandard\CommandLine\Environment\Environment;
+
+use function Safe\file_get_contents;
+use function Safe\scandir;
+
+class CheckSafeFunctionUsageSniff implements Sniff
+{
+    /** @var array<string> */
+    private array $functionNames;
+
+    public function __construct()
+    {
+        $container = ContainerFactory::getContainerInstance();
+        $environment = $container->get(Environment::class);
+        $path = $environment->getRootDirectory()->getRealPath() . '/vendor/thecodingmachine/safe/generated/';
+        try {
+            // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
+            $filesUnfiltered = @scandir($path);
+        } catch (DirException $exception) {
+            return;
+        }
+
+        $files = array_diff($filesUnfiltered, ['.', '..', 'Exceptions']);
+
+        $parser = $container->get(Parser::class);
+
+        $functionNames = [];
+
+        foreach ($files as $file) {
+            $ast = $parser->parse(file_get_contents($path . $file));
+            $nodeFinder = $container->get(NodeFinder::class);
+
+            $functions = $nodeFinder->find($ast, static fn(Node $node) => $node instanceof Node\Stmt\Function_);
+            $functionNamesLocal = array_map(static fn(Node\Stmt\Function_ $node) => (string) $node->name, $functions);
+
+            $functionNames = [...$functionNames, ...$functionNamesLocal];
+        }
+
+        $this->functionNames = $functionNames;
+    }
+
+    /**
+     * Description is in the inherited doc.
+     *
+     * @{inheritDoc}
+     *
+     * @phpcsSuppress SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingTraversableTypeHintSpecification We must
+     *                stay compatible with the interface even if we don't like it.
+     */
+    public function register(): array
+    {
+        return [
+            T_STRING,
+        ];
+    }
+
+    /**
+     * Description is in the inherited doc.
+     *
+     * @{inheritDoc}
+     *
+     * @phpcsSuppress SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint We must
+     *                stay compatible with the interface even if we don't like it.
+     *
+     * @throws RuntimeException
+     */
+    public function process(File $phpcsFile, $stackPtr): void
+    {
+        if (!isset($this->functionNames)) {
+            throw new RuntimeException(
+                'No function names found! Did you forget to install thecodingmachine/Safe?',
+                1684240278
+            );
+        }
+
+        $tokens = $phpcsFile->getTokens();
+        $functionName = strtolower(ltrim($tokens[$stackPtr]['content'], '\\'));
+
+        try {
+            $this->assertNextTokenParenthesisOpener($phpcsFile, $stackPtr);
+            $this->assertGlobaleFunctionCall($phpcsFile, $stackPtr);
+            $this->assertFunctionProvidedBySafe($tokens[$stackPtr]['content']);
+            $this->assertFunctionUnused($phpcsFile, $functionName);
+        } catch (RuntimeException $exception) {
```

Comment:
vielleicht lieber nen eigenen exception type werfen und fangen? hab keinen überblick, was in den helper calls so alles passiert, aber undenkbar isses nicht, das irgendwann irgendwp ne runtimeexception aus anderen gründen fliegt

Reply:
Verdammt. Ich hatte befürchtet, dass das passiert. Das ist ein eigener Exception Type (liegt im selben Namespace daher kein use.), aber die Namen sind zu ähnlich. Ich denk mir was anderes aus.